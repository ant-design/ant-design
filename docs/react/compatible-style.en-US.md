---
group:
  title: Advanced
order: 1
title: CSS Compatible
---

### Default Style Compatibility

Ant Design supports the [last 2 versions of modern browsers](https://browsersl.ist/#q=defaults). If you need to be compatible with legacy browsers, please perform downgrade processing according to actual needs:

| Feature | antd version | Compatibility | Minimum Chrome Version | Compatibility workaround |
| --- | --- | --- | --- | --- |
| [:where Selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:where) | `>=5.0.0` | [caniuse](https://caniuse.com/?search=%3Awhere) | Chrome 88 | `<StyleProvider hashPriority="high">` |
| [CSS Logical Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties) | `>=5.0.0` | [caniuse](https://caniuse.com/css-logical-props) | Chrome 89 | `<StyleProvider transformers={[legacyLogicalPropertiesTransformer]}>` |

If you need to support older browsers, please use [StyleProvider](https://github.com/ant-design/cssinjs#styleprovider) for degradation handling according to your actual requirements.

## `:where` in selector

- antd version: `>=5.0.0`
- MDN: [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/:where)
- Browser Compatibility: [caniuse](https://caniuse.com/?search=%3Awhere)
- Minimum Chrome Version Supported: 88
- Default Enabled: Yes

The CSS-in-JS feature of Ant Design uses the ":where" selector by default to lower the CSS selector specificity, reducing the additional cost of adjusting custom styles when upgrading for users. However, the compatibility of the ":where" syntax is relatively poor in older browsers ([compatibility](https://developer.mozilla.org/en-US/docs/Web/CSS/:where#browser_compatibility)). In certain scenarios, if you need to support older browsers, you can use `@ant-design/cssinjs` to disable the default lowering of specificity (please ensure version consistency with antd).

```tsx
import { StyleProvider } from '@ant-design/cssinjs';

// Config `hashPriority` to `high` instead of default `low`
// Which will remove `:where` wrapper
export default () => (
  <StyleProvider hashPriority="high">
    <MyApp />
  </StyleProvider>
);
```

It will turn `:where` to class selector:

```diff
--  :where(.css-bAMboO).ant-btn {
++  .css-bAMboO.ant-btn {
      color: #fff;
    }
```

Note: After turning off the `:where` downgrade, you may need to manually adjust the priority of some styles. Or you can **use PostCSS plugin** to raise application css selector priority. PostCSS provides many plugins can help on this. e.g:

- [postcss-scopify](https://www.npmjs.com/package/postcss-scopify)
- [postcss-increase-specificity](https://www.npmjs.com/package/postcss-increase-specificity)
- [postcss-add-root-selector](https://www.npmjs.com/package/postcss-add-root-selector)

Raise priority through plugin:

```diff
--  .my-btn {
++  #root .my-btn {
      background: red;
    }
```

## CSS Logical Properties

- antd version: `>=5.0.0`
- MDN：[CSS Logical Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties)
- Browser Compatibility: [caniuse](https://caniuse.com/css-logical-props)
- Minimum Chrome Version Supported: 89
- Default Enabled: Yes

To unify LTR and RTL styles, Ant Design uses CSS logical properties. For example, the original `margin-left` is replaced by `margin-inline-start`, so that it is the starting position spacing under both LTR and RTL. If you need to be compatible with older browsers, you can configure `transformers` through the `StyleProvider` of `@ant-design/cssinjs`:

```tsx
import { legacyLogicalPropertiesTransformer, StyleProvider } from '@ant-design/cssinjs';

// `transformers` provides a way to transform CSS properties
export default () => (
  <StyleProvider transformers={[legacyLogicalPropertiesTransformer]}>
    <MyApp />
  </StyleProvider>
);
```

When toggled, styles will downgrade CSS logical properties:

```diff
.ant-modal-root {
-- inset: 0;
++ top: 0;
++ right: 0;
++ bottom: 0;
++ left: 0;
}
```

## `@layer`

- antd version: `>=5.17.0`
- MDN：[CSS @layer](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer)
- Browser Compatibility: [caniuse](https://caniuse.com/css-at-rule-layer)
- Minimum Chrome Version Supported: 99
- Default Enabled: No

Ant Design supports configuring `@layer` for unified css priority downgrade since `5.17.0`. After the downgrade, the style of antd will always be lower than the default CSS selector priority, so that users can override the style (please be sure to check the browser compatibility of `@layer`):

```tsx
import { StyleProvider } from '@ant-design/cssinjs';

export default () => (
  <StyleProvider layer>
    <MyApp />
  </StyleProvider>
);
```

antd styles will be encapsulated in `@layer` to lower the priority:

```diff
++  @layer antd {
      :where(.css-bAMboO).ant-btn {
        color: #fff;
      }
++  }
```

## Rem Adaptation

In responsive web development, there is a need for a convenient and flexible way to achieve page adaptation and responsive design. The `px2remTransformer` transformer can quickly and accurately convert pixel units in style sheets to rem units relative to the root element (HTML tag), enabling the implementation of adaptive and responsive layouts.

```tsx
import { px2remTransformer, StyleProvider } from '@ant-design/cssinjs';

const px2rem = px2remTransformer({
  rootValue: 32, // 32px = 1rem; @default 16
});

export default () => (
  <StyleProvider transformers={[px2rem]}>
    <MyApp />
  </StyleProvider>
);
```

The resulting transformed styles:

```diff
 .px2rem-box {
-  width: 400px;
+  width: 12.5rem;
   background-color: green;
-  font-size: 32px;
+  font-size: 1rem;
   border: 10PX solid #f0f;
 }

 @media only screen and (max-width: 600px) {
   .px2rem-box {
     background-color: red;
-    margin: 10px;
+    margin: 0.3125rem;
   }
 }
```

### Options

<!-- prettier-ignore -->
| Parameter | Description  | Type | Default |
| --- | --- | --- | --- |
| rootValue | Font size of the root element | `number` | 16 |
| precision | Decimal places for the converted value | `number` | 5 |
| mediaQuery | Whether to convert px in media queries | `boolean` | false |

For more details, please refer to: [px2rem.ts#Options](https://github.com/ant-design/cssinjs/blob/master/src/transformers/px2rem.ts)

## Shadow DOM Usage

Since `<style />` tag insertion is different from normal DOM in Shadow DOM scenario, you need to use `StyleProvider` of `@ant-design/cssinjs` to configure the `container` property to set the insertion position:

```tsx
import { StyleProvider } from '@ant-design/cssinjs';
import { createRoot } from 'react-dom/client';

const shadowRoot = someEle.attachShadow({ mode: 'open' });
const container = document.createElement('div');
shadowRoot.appendChild(container);
const root = createRoot(container);

root.render(
  <StyleProvider container={shadowRoot}>
    <MyApp />
  </StyleProvider>,
);
```

## Compatible with Third-party Style Libraries

In some cases, you may need antd to coexist with other style libraries, such as `Tailwind CSS`, `Emotion`, `styled-components`, etc. Unlike traditional CSS solutions, these third-party libraries are often not easy to override antd styles by increasing CSS selector priority. You can configure `@layer` for antd to lower its CSS selector weight, and arrange `@layer` order to solve style override problems:

### antd config `@layer`

```tsx
import { StyleProvider } from '@ant-design/cssinjs';

export default () => (
  <StyleProvider layer>
    <MyApp />
  </StyleProvider>
);
```

### TailwindCSS Arrange `@layer`

In global.css, adjust `@layer` to control the order of style override. Place `tailwind-base` before `antd`:

```less
@layer tailwind-base, antd;

@layer tailwind-base {
  @tailwind base;
}
@tailwind components;
@tailwind utilities;
```

### reset.css

If you use antd's `reset.css` style, you need to specify `@layer` for it to prevent the style from overriding antd:

```less
@layer reset, antd;

@import url(reset.css) layer(reset);
```

### With other CSS-in-JS libraries

After configuring `@layer` for antd, you don't need to do any additional configuration for other CSS-in-JS libraries. Your CSS-in-JS can completely override antd styles.

### SSR Scene

When using SSR, styles are often rendered inline in HTML through `<style />`. At this time, please make sure that the styles with the specified `@layer` priority order are loaded before `@layer` is used.

#### ❌ Wrong

```html
<head>
  <!-- SSR Injection style -->
  <style>
    @layer antd {
      /** ... */
    }
  </style>

  <!-- css file contains @layer xxx, antd; -->
  <link rel="stylesheet" href="/b9a0m0b9o0o3.css" />
  <!-- or write @layer xxx, antd; in html directly -->
  <style>
    @layer xxx, antd;
  </style>
</head>
```

#### ✅ Correct

```html
<head>
  <!-- css file contains @layer xxx, antd; -->
  <link rel="stylesheet" href="/b9a0m0b9o0o3.css" />
  <!-- or write @layer xxx, antd; in html directly -->
  <style>
    @layer xxx, antd;
  </style>

  <!-- SSR Injection style -->
  <style>
    @layer antd {
      /** ... */
    }
  </style>
</head>
```
