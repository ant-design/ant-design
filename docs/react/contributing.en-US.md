---
group:
  title: Other
order: 1
title: Contributing
---

The following is a set of guidelines for contributing to Ant Design. Please spend several minutes reading these guidelines before you create an issue or pull request.

## Code of Conduct

We have adopted a [Code of Conduct](https://github.com/ant-design/ant-design/blob/master/CODE_OF_CONDUCT.md) that we expect project participants to adhere to. Please read the full text so that you can understand what actions will and will not be tolerated.

## Open Development

All work on Ant Design happens directly on [GitHub](https://github.com/ant-design). Both core team members and external contributors send pull requests which go through the same review process.

## Branch Organization

According to our [release schedule](/changelog#release-schedule), we maintain two branches, `master` and `feature`. If you send a bugfix pull request, please do it against the `master` branch, if it's a feature pull request, please do it against the `feature` branch.

## Bugs

We are using [GitHub Issues](https://github.com/ant-design/ant-design/issues) for bug tracking. The best way to get your bug fixed is by using our [issue helper](http://new-issue.ant.design) and provide reproduction steps with this [template](https://u.ant.design/codesandbox-repro).

Before you report a bug, please make sure you've searched existing issues, and read our [FAQ](/docs/react/faq).

## Proposing a Change

If you intend to change the public API or introduce new feature, we also recommend you use our [issue helper](http://new-issue.ant.design) to create a feature request issue.

If you want to help on new API, please reference [API Naming Rules](https://github.com/ant-design/ant-design/wiki/API-Naming-rules) to name it.

## Your First Pull Request

Working on your first Pull Request? You can learn how from this free video series:

[How to Contribute to an Open Source Project on GitHub](https://egghead.io/courses/how-to-contribute-to-an-open-source-project-on-github)

To help you get your feet wet and get you familiar with our contribution process, we have a list of [good first issues](https://github.com/ant-design/ant-design/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) that contain bugs or small features that have a relatively limited scope. This is a great place to get started.

If you decide to fix an issue, please be sure to check the comment thread in case somebody is already working on a fix. If nobody is working on it at the moment, please leave a comment stating that you intend to work on it so other people don't accidentally duplicate your effort.

If somebody claims an issue but doesn't follow up for more than two weeks, it's fine to take over it but you should still leave a comment.

## Sending a Pull Request

The core team is monitoring for pull requests. We will review your pull request and either merge it, request changes to it, or close it with an explanation.

**Before submitting a pull request**, please make sure the following is done:

1. Fork the repository and create your branch from the [correct branch](#branch-organization).
2. Run `npm install` in the repository root.
3. If you've fixed a bug or added code that should be tested, add tests!
4. Ensure the test suite passes (npm run test). Tip: `npm test -- --watch TestName` is helpful in development.
5. Run `npm test -- -u` to update the [jest snapshots](https://jestjs.io/docs/snapshot-testing) and commit these changes as well (if there are any updates).
6. Ensure the UI change passes `npm run test:image`, Run `npm run test:image -- -u` to update UI snapshots and commit these changes as well (if there are any updates), **UI test base on [Docker](https://docs.docker.com/get-docker/), need download the corresponding installation according to the platform**
7. Make sure your code lints (npm run lint). Tip: Lint runs automatically when you `git commit` (Use [Git Hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)).
8. Finally, please make sure that all GitHub CI checks pass, if they fail, you can click `detail` to enter the details to view the reason.

Sending a Pull Request to [react-component](https://github.com/react-component/):

Since antd's components are based on react-component, sometimes you may need to send pull request to the corresponding react-component repository. If it's a bugfix pull request, after it's merged, the core team will release a patch release for that component as soon as possible, then you only need to reinstall antd in your project to get the latest patch release. If it's a feature pull request, after it's merged, the core team will release a minor release, then you need raise another pull request to [Ant Design](https://github.com/ant-design/ant-design/) to update dependencies, document and TypeScript interfaces (if needed).

## Development Workflow

`npm` or `yarn` are recommended as package management tools.

After you clone the antd code and use following commands to install dependencies:

<InstallDependencies npm='$ npm install' yarn='$ yarn'></InstallDependencies>

You can also run the following common commands:

### Run locally

Runs Ant Design website locally.

<InstallDependencies npm='$ npm start' yarn='$ yarn start'></InstallDependencies>

### Checks the code style

<InstallDependencies npm='$ npm run lint' yarn='$ yarn lint'></InstallDependencies>

### Run test

runs the complete test suite. (Make sure the `NODE_ENV` environment variable is unset, or it may causing some problems.)

<InstallDependencies npm='$ npm test' yarn='$ yarn test'></InstallDependencies>

### Compile

compiles TypeScript code to the `lib` and `es` directory.

<InstallDependencies npm='$ npm run compile' yarn='$ yarn compile'></InstallDependencies>

### Build

creates UMD build of antd.

<InstallDependencies npm='$ npm run dist' yarn='$ yarn dist'></InstallDependencies>

## Development Tools

- VSCode plugin for CSS in JS: https://marketplace.visualstudio.com/items?itemName=shezhangzhang.antd-design-token
- Ant Design cheatsheet in VS Code: https://github.com/fi3ework/vscode-antd-rush

## Being a collaborator

If you are an active contributor and are willing to work with Ant Design Team in our opensource workflow, you can [apply to be a outside collaborator](https://github.com/ant-design/ant-design/wiki/Collaborators#how-to-apply-for-being-a-collaborator).

You can also refer to the following contribution guide to become an antd contributor:

- [How to Grow as a Collaborator](/docs/blog/to-be-collaborator)
