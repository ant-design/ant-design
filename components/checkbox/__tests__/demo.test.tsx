import * as React from 'react';

import demoTest, { rootPropsTest } from '../../../tests/shared/demoTest';

demoTest('checkbox');

rootPropsTest(
  'checkbox',
  (Checkbox, props) => <Checkbox.Group {...props} value={[]} options={['Bamboo']} />,
  {
    name: 'Checkbox.Group',
  },
);
