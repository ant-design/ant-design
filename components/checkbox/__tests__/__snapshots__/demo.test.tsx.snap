// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders components/checkbox/demo/basic.tsx correctly 1`] = `
<div
  class="ant-checkbox-wrapper"
>
  <span
    class="ant-checkbox ant-wave-target"
  >
    <input
      class="ant-checkbox-input"
      type="checkbox"
    />
    <span
      class="ant-checkbox-inner"
    />
  </span>
  <span>
    Checkbox
  </span>
</div>
`;

exports[`renders components/checkbox/demo/check-all.tsx correctly 1`] = `
Array [
  <div
    class="ant-checkbox-wrapper"
  >
    <span
      class="ant-checkbox ant-checkbox-indeterminate ant-wave-target"
    >
      <input
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
    <span>
      Check all
    </span>
  </div>,
  <div
    class="ant-divider ant-divider-horizontal"
    role="separator"
  />,
  <div
    class="ant-checkbox-group"
  >
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-checked"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          type="checkbox"
          value="Apple"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Apple
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Pear"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pear
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-checked"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          type="checkbox"
          value="Orange"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Orange
      </span>
    </div>
  </div>,
]
`;

exports[`renders components/checkbox/demo/controller.tsx correctly 1`] = `
Array [
  <p
    style="margin-bottom:20px"
  />,
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-checked"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-checked"
    >
      <input
        checked=""
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
    <span>
      Checked-Enabled
    </span>
  </div>,
  <p />,
  <p>
    <button
      class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-sm"
      type="button"
    >
      <span>
        Uncheck
      </span>
    </button>
    <button
      class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-sm"
      style="margin:0 10px"
      type="button"
    >
      <span>
        Disable
      </span>
    </button>
  </p>,
]
`;

exports[`renders components/checkbox/demo/custom-line-width.tsx correctly 1`] = `
Array [
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-checked"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-checked"
    >
      <input
        checked=""
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
  <div
    class="ant-checkbox-wrapper"
  >
    <span
      class="ant-checkbox ant-wave-target"
    >
      <input
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-checked"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-checked"
    >
      <input
        checked=""
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
  <div
    class="ant-checkbox-wrapper"
  >
    <span
      class="ant-checkbox ant-wave-target"
    >
      <input
        class="ant-checkbox-input"
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
]
`;

exports[`renders components/checkbox/demo/debug-disable-popover.tsx correctly 1`] = `
<div
  style="padding:56px"
>
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-wrapper-disabled"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-checked ant-checkbox-disabled"
    >
      <input
        checked=""
        class="ant-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>
</div>
`;

exports[`renders components/checkbox/demo/debug-line.tsx correctly 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
    style="border:1px solid red;margin-bottom:16px"
  >
    <div
      class="ant-space-item"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="light"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <div>
        Bamboo
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="little"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Little
        </span>
      </div>
    </div>
  </div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
    style="border:1px solid red;margin-bottom:16px"
  >
    <div
      class="ant-space-item"
    >
      <label
        class="ant-radio-wrapper"
      >
        <span
          class="ant-radio ant-wave-target"
        >
          <input
            class="ant-radio-input"
            type="radio"
            value="light"
          />
          <span
            class="ant-radio-inner"
          />
        </span>
      </label>
    </div>
    <div
      class="ant-space-item"
    >
      <div>
        Bamboo
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <label
        class="ant-radio-wrapper"
      >
        <span
          class="ant-radio ant-wave-target"
        >
          <input
            class="ant-radio-input"
            type="radio"
            value="little"
          />
          <span
            class="ant-radio-inner"
          />
        </span>
        <span>
          Little
        </span>
      </label>
    </div>
  </div>
  <div
    style="border:1px solid red;margin-bottom:16px;display:flex;align-items:center"
  >
    <div
      class="ant-checkbox-wrapper"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="light"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
    </div>
    <div>
      Bamboo
    </div>
    <div
      class="ant-checkbox-wrapper"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="little"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Little
      </span>
    </div>
  </div>
  <div
    style="border:1px solid red;margin-bottom:16px;display:flex;align-items:center"
  >
    <label
      class="ant-radio-wrapper"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="light"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
    </label>
    <div>
      Bamboo
    </div>
    <label
      class="ant-radio-wrapper"
    >
      <span
        class="ant-radio ant-wave-target"
      >
        <input
          class="ant-radio-input"
          type="radio"
          value="little"
        />
        <span
          class="ant-radio-inner"
        />
      </span>
      <span>
        Little
      </span>
    </label>
  </div>
  <div>
    <div
      class="ant-checkbox-wrapper"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Aligned
      </span>
    </div>
  </div>
  <div>
    <div
      class="ant-checkbox-wrapper"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        <span
          style="font-size:32px"
        >
          Aligned
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`renders components/checkbox/demo/disabled.tsx correctly 1`] = `
Array [
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-disabled"
    >
      <input
        class="ant-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
  <br />,
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
  >
    <span
      class="ant-checkbox ant-checkbox-indeterminate ant-wave-target ant-checkbox-disabled"
    >
      <input
        class="ant-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
  <br />,
  <div
    class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
  >
    <span
      class="ant-checkbox ant-wave-target ant-checkbox-checked ant-checkbox-disabled"
    >
      <input
        checked=""
        class="ant-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="ant-checkbox-inner"
      />
    </span>
  </div>,
]
`;

exports[`renders components/checkbox/demo/group.tsx correctly 1`] = `
Array [
  <div
    class="ant-checkbox-group"
  >
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-checked"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          type="checkbox"
          value="Apple"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Apple
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Pear"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pear
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Orange"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Orange
      </span>
    </div>
  </div>,
  <br />,
  <br />,
  <div
    class="ant-checkbox-group"
  >
    <div
      class="ant-checkbox-wrapper ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Apple"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Apple
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-checked"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          type="checkbox"
          value="Pear"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pear
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Orange"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Orange
      </span>
    </div>
  </div>,
  <br />,
  <br />,
  <div
    class="ant-checkbox-group"
  >
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked ant-checkbox-wrapper-disabled ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-checked ant-checkbox-disabled"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          disabled=""
          type="checkbox"
          value="Apple"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Apple
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-disabled"
      >
        <input
          class="ant-checkbox-input"
          disabled=""
          type="checkbox"
          value="Pear"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pear
      </span>
    </div>
    <div
      class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled ant-checkbox-group-item"
    >
      <span
        class="ant-checkbox ant-wave-target ant-checkbox-disabled"
      >
        <input
          class="ant-checkbox-input"
          disabled=""
          type="checkbox"
          value="Orange"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Orange
      </span>
    </div>
  </div>,
]
`;

exports[`renders components/checkbox/demo/layout.tsx correctly 1`] = `
<div
  class="ant-checkbox-group"
  style="width:100%"
>
  <div
    class="ant-row"
  >
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="A"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          A
        </span>
      </div>
    </div>
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="B"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          B
        </span>
      </div>
    </div>
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="C"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          C
        </span>
      </div>
    </div>
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="D"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          D
        </span>
      </div>
    </div>
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-checkbox-wrapper"
      >
        <span
          class="ant-checkbox ant-wave-target"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
            value="E"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          E
        </span>
      </div>
    </div>
  </div>
</div>
`;
